### DISCUSSION TITLE
What plugins do you need to use swc in your app?

### DISCUSSION THREAD

**Discussion by kwonoj** [swc-project/swc#4768] [Ideas]
### Context

There are numbers of request to add features in `@swc/core`. It is currently scattered around gh issues and discussions, and not clearly compiled as list as some are requested as core features, and some goes to the plugins and some elses.

This post attempts to collect list of the features people use, but not available in swc currently which blocks adaption of the swc in the certain app user currently build.

Few disclaimer:

- **THIS DOES NOT GUARANTEE SWC WILL IMPLEMENT CERTAIN FEATURE / OR CREATE PLUGINS**. 
   This is more of collecting real world usecases for now.
- Please check https://github.com/swc-project/plugins just in case if some of features are available already.
- Please do not leave duplicated comments (or +1, something similar) if there's existing comment. Instead, please use emoji to express interests.
- Please share link to original plugin you're using currently, either babel, or next-swc, or something else.
- Please scope discussion to the features need to be implemented as plugin. Generic plugin discussion should be discussed in https://github.com/swc-project/swc/discussions/3540

There are some existing plugins. Check https://github.com/swc-project/swc/discussions/4818 for the details.

#### NOTE:

**Any plugin relies on `babel-plugin-macro` have same constraint to https://github.com/swc-project/swc/discussions/4768#discussioncomment-3436424 (`babel-plugin-macros pass babel object to macros`) - which SWC can't provide feature parity to macro itself. We do not have any immediate plan to investigate similar approaches. **

**Comment by XiNiHa** [comment]
SolidJS/dom-expressions JSX transform, although it seems like some of the SolidJS community members are already giving it a shot.

**Reply by orionmiz** [reply]
working on this plugin :)

**Reply by Keats** [reply]
Any news on that?

**Reply by XiNiHa** [reply]
I'm also currently contributing to the plugin and made some progress - although it doesn't support some critical features like `ref` and SSR yet.
https://github.com/modderme123/swc-plugin-jsx-dom-expressions

**Comment by natew** [comment]
At least for Tamagui, the ideal would be a swc-babel-api that mimicked babel as much as possible. Not sure if the fundamentals are so different that it’s impossible, but honestly the Babel API is incredibly good and all the useful helpers there make a big difference.

**Reply by kwonoj** [reply]
This is not for specific features to be implemented via plugin, right? Also yes, there are fundamental difference between babel to swc and its runtime (rust) that we'll never provide 1:1 api surface to babel does.

Double clarify, we should provide helpers - just saying some things are not possible such as access / mutate parent nodes, etcs.

**Reply by kwonoj** [reply]
Please use https://github.com/swc-project/swc/discussions/3540 for generic plugin discussions.

**Comment by Edge00** [comment]
Hope swc can prodive capacity like [@vue/babel-plugin-jsx](https://github.com/vuejs/babel-plugin-jsx) to handle `.vue` written in `.tsx`.

**Reply by XiNiHa** [reply]
Although I personally do not use Vue anymore, I really wanted to have this while using Vue. Will be a huge one I guess.

**Reply by Edge00** [reply]
> not only vue jsx, but vue sfc plugin that can support script-setup

You mean using `swc` as a bundler? I'am using webpack with `vue-loader@16` and `swc-loader`, which can work well with `script-setup`.
However, using swc to derectly compile `.vue` file is also a great idea!

**Reply by Edge00** [reply]
That would be awsome! If so, the [sfc playground](https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8aDE+e3sgbXNnIH19PC9oMT5cbiAgPENvbXAgbmFtZT1cIkNoaWxkXCIgLz5cbjwvdGVtcGxhdGU+XG5cbjxzY3JpcHQgc2V0dXA+XG5pbXBvcnQgQ29tcCBmcm9tICcuL0NvbXAudnVlJ1xuY29uc3QgbXNnID0gJ0hlbGxvIFNldHVwIFNjcmlwdCEnXG48L3NjcmlwdD4iLCJDb21wLnZ1ZSI6Ijx0ZW1wbGF0ZT5cbiAgPGgxPnt7IG1zZyB9fTwvaDE+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IHNldHVwPlxuY29uc3QgbXNnID0gJ0NoaWxkJ1xuPC9zY3JpcHQ+IiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge1xuICAgIFwidnVlXCI6IFwiaHR0cHM6Ly9zZmMudnVlanMub3JnL3Z1ZS5ydW50aW1lLmVzbS1icm93c2VyLmpzXCJcbiAgfVxufSJ9) could be compiled by rust.

**Reply by g-plane** [reply]
I made it: [g-plane/swc-plugin-vue-jsx](https://github.com/g-plane/swc-plugin-vue-jsx).

**Comment by mwojick** [comment]
Would like the Emotion plugin to get separated out of Next: https://github.com/vercel/next.js/pull/34687
We can't move to Next but would like to at least swap babel with swc for now. Emotion transform is the only blocker.

**Reply by kwonoj** [reply]
https://github.com/swc-project/plugins/pull/29

**Reply by IvanRodriCalleja** [reply]
@mwojick You can check. https://github.com/IvanRodriCalleja/emotion-swc-plugin

**Comment by joooye34** [comment]
Feedback from TikTok Web team:

- [ ] Loadable components transform plugin (https://github.com/gregberge/loadable-components/tree/main/packages/babel-plugin)
- [x] emotion-ts-plugin
- [ ] ts-import-plugin (https://github.com/Brooooooklyn/ts-import-plugin)

**Reply by kdy1** [reply]
Loadable components is implemented by https://github.com/swc-project/plugins/pull/76

**Reply by kdy1** [reply]
For `ts-import-plugin`, take a look at https://github.com/swc-project/plugins/tree/28ab1c04df68f7347d470ed3bac4e72d7f7befa9/packages/transform-imports

**Comment by kwonoj** [comment]
`react-native-reanimdated` babel plugin

- https://docs.expo.dev/versions/latest/sdk/reanimated (https://github.com/software-mansion/react-native-reanimated/blob/main/plugin.js)

**Reply by kwonoj** [reply]
https://github.com/software-mansion/react-native-reanimated/pull/3341

**Comment by jaw-sh** [comment]
Throwing my hat in for SASS/SCSS support, as mentioned in discussion.

I'm writing an Actix web application and it'd be nice to stay in the Rust ecosystem for asset compilation. It would provide interesting options for the application to manage its own assets directly.

**Comment by PawelPyc154** [comment]
[babel-plugin-macros](https://github.com/kentcdodds/babel-plugin-macros)

**Reply by Fumler** [reply]
Same here, and to add some context, we're using [twin.macro](https://github.com/ben-rogerson/twin.macro) through babel-plugin-macros

**Reply by kdy1** [reply]
`babel-plugin-macros` pass `babel` object to macros, so it's not possible to port it without modifying macro libraries

**Reply by watsonhaw5566** [reply]
> `babel-plugin-macros` pass `babel` object to macros, so it's not possible to port it without modifying macro libraries

Can i use another way to implement [babel-plugin-macros](https://github.com/kentcdodds/babel-plugin-macros) on swc ?

**Comment by vip30** [comment]
https://github.com/davesnx/babel-plugin-transform-react-qa-classes

**Comment by Dremora** [comment]
[@vanilla-extract/babel-plugin](https://github.com/seek-oss/vanilla-extract)
While vanilla-extract works fine with swc and without babel plugin, the plugin is required [for testing](https://vanilla-extract.style/documentation/test-environments/), making it impossible to test Next.js application using vanilla-extract.

**Reply by kwonoj** [reply]
Marking as resolved. https://github.com/seek-oss/vanilla-extract/discussions/478#discussioncomment-3640551

**Comment by ild0tt0re** [comment]
To use Storybook with complete SWC support, the [**storybook-addon-swc** plugin](https://storybook.js.org/addons/storybook-addon-swc) is not enough. 
We would like to have:
- [babel-plugin-named-exports-order](https://github.com/storybookjs/babel-plugin-named-exports-order)
- [babel-plugin-react-docgen](https://github.com/storybookjs/babel-plugin-react-docgen)

**Comment by belaczek** [comment]
We're currently blocked by dependency on [babel-plugin-macros](https://github.com/kentcdodds/babel-plugin-macros) by [LinguiJs](https://github.com/lingui/js-lingui) library. Having a custom swc plugin for Lingui would enable us to switch from babel.

**Reply by kwonoj** [reply]
https://github.com/swc-project/swc/discussions/4768#discussioncomment-3436424

**Reply by renchap** [reply]
@kwonoj Babel Macros can not directly be ported, but a Lingui plugin could maybe replace the need for macros?

We are blocked by Lingui as well, and switching to a new I18n solution is a huge task…

**Reply by kwonoj** [reply]
I wouldn't call it 100% impossible task, but that means someone basically needs to reinvent whole logics existing plugin relying on macro / babel specifics. It is not trivial task to consider.

**Reply by timofei-iatsenko** [reply]
I've started implementation here https://github.com/thekip/swc-lingui-plugin

Probably need few more weeks to get this done. 

Also would be very happy if someone from swc core team could review my code for performance and general plugin tips.

**Reply by timofei-iatsenko** [reply]
Resolved: https://github.com/lingui/swc-plugin

**Comment by tfar** [comment]


**Reply by kwonoj** [reply]
Webpack loaders are irrelevant to swc.

**Comment by xyhxx** [comment]
We need [babel-plugin-react-remove-properties](https://github.com/oliviertassinari/babel-plugin-react-remove-properties), it's completed in [next.js](https://github.com/vercel/next.js/pull/31606).

**Reply by g-plane** [reply]
I made a SWC plugin "swc-plugin-react-remove-properties": [GitHub repo](https://github.com/g-plane/swc-plugin-react-remove-properties), [npm](https://www.npmjs.com/package/swc-plugin-react-remove-properties).

**Comment by meskill** [comment]
[@babel/plugin-transform-react-constant-elements](https://babeljs.io/docs/en/babel-plugin-transform-react-constant-elements) to optmize bundle for the react app

**Reply by JSerFeng** [reply]
Working on it

**Reply by JSerFeng** [reply]
Take a look at here https://github.com/swc-project/swc/discussions/7830#discussioncomment-6757618

**Comment by mmichalskiba** [comment]
[babel-plugin-transform-goober](https://github.com/cristianbote/goober/tree/master/packages/babel-plugin-transform-goober) to rewrite code from `styled.div` to `styled('div')` would be much appreciated!

**Comment by erkanaltinors** [comment]
[babel-plugin-jsx-control-statements](https://www.npmjs.com/package/babel-plugin-jsx-control-statements)

**Comment by hjkcai** [comment]
I made some utility plugins
- https://github.com/hjkcai/swc-plugin-add-display-name
- https://github.com/hjkcai/swc-plugin-enum-to-obj

**Comment by AmalJossy** [comment]
https://github.com/gajus/babel-plugin-react-css-modules

**Comment by ohsory1324** [comment]
https://github.com/facebook/stylex/tree/main/packages/babel-plugin

**Comment by philwolstenholme** [comment]
Something like https://github.com/lemonmade/babel-plugin-react-component-data-attribute or https://github.com/davesnx/babel-plugin-transform-react-qa-classes would be great. 

https://github.com/hjkcai/swc-plugin-add-display-name is close but sets a display name for React Devtools, whereas those two plugins set a data attribute with the name (for testing or for marketing folk to use when setting up click tracking etc)

**Comment by msdrigg** [comment]
It would be helpful to have something like this:

`plugin-transform-react-jsx-source` https://babeljs.io/docs/babel-plugin-transform-react-jsx-source

This makes react errors show up better

**Comment by gethari** [comment]
anything for `lit-element` or `lit` component lib app ?
